### PR TITLE
Espressobin firmware

### PR DESCRIPTION
**Pull request by ManoftheSea**
###### Description of changes

Update uboot package to build for the Espressobin SBC

Add mox-boot-builder package to build a trusted environment program for Marvell a3k SOCs

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Platform firmware for this Marvell SOC seems to build backward from the Rockchip method.  Here, u-boot is built without referencing arm-trusted-firmware, and then armTrustedFirmwareA3700 would bring u-boot.bin, wtmi_app.bin, and a few other files that I haven't understood how to bring into Nix.  U-boot is bumped to 2023.01 because there are specific fixes for espressobin that aren't in earlier versions.

### CODE REVIEW

**Comment by NickCao** [comment]
We currently have a policy not to add vendor kernels to nixpkgs, donno if that applies to uboot as well. But usually these hardware specific bits would be a better fit for https://github.com/NixOS/nixos-hardware.

**Comment by lopsided98** [comment]
I think most of this PR would be acceptable in a more complete form. It appears to use upstream U-Boot (although we should update U-Boot generally rather than pinning a specific version for this device), and assuming it uses upstream ATF, that should be fine as well. I'm not sure how `mox-boot-builder` fits into everything, but it looks like it fills a similar role to tools we already include for Amlogic devices.

@samueldr is more concerned about vendor specific code than I am though, so it's really his call.

**Comment by ManoftheSea** [comment]
I would like to see uboot update and not pin for espressobin.  But that's a bit of a higher level call than I can make.
Upstream ATF also applies, but the build process is backward from the rockchips: instead of the RK process of building ATF and including it in u-boot, the Marvell chip uses u-boot's bin as an input, along with wtmi_app from Mox, two more marvell git repos, and the ATF code.
The Mox code is *probably* free software, but I can't find a license on their repo.  The repo is intended for the Mox Turris, but the wtmi is useful for the espressobin (these share the SOC). 

To be truly useful, there is additional packaging work (that I hope to do) for armTrustedFirmwareA3700 and maybe the mv-ddr-marvell (gpl2+) and a3700-utils-marvell (bsd 3-clause) repos.

File: pkgs/misc/uboot/default.nix
Comment (samueldr) [inline pkgs/misc/uboot/default.nix:205]: Like @lopsided98 noticed, we should be updating U-Boot first.

We should probably block this PR until our U-Boot build is updated for 2023.01. And then remove this.

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:43]: We'll need a maintainer here, normally it should be you.

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:44]: This should be the target platform of the resulting output; same as for U-Boot:

```suggestion
    platforms = ["aarch64-linux"];
```

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:42]: Do we have details or citaiton on it being redistributable? It's *probably fine*, but ideally if we can cite it'd be better.

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:33]:
```suggestion
    mkdir -p $out
    cp wtmi_app.bin $out
```

We don't neeed the level of indirection here. `$out` is the proper way to refer to the (default) output. We don't have enough complexity to warrant concating this way.

> **TIP** in this case `lib.escapeShellArgs` would have been preferred, as it would correctly handle complex cases.

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:27]:
```suggestion
```

(See following comment)

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:18]:
```suggestion
```

(See following comment)

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:16]: When given a gitlab source, we should use `fetchFromGitLab`. It uses the archive downloads rather than doing a more costly git clone. You can point to a non-`gitlab.com` host with the `domain` arg.

Note that this may (will?) change the hash, so make sure to update it as you would usually.

**Review by samueldr — CHANGES_REQUESTED** [review]
First pass at reading/reviewing. I'll need to read more about how upstream U-Boot expects this all to be built, and see if this is in line.

**Comment by samueldr** [comment]
Converted to draft, as AFAIUI this is incomplete, do tell if I got this wrong. I'm understanding from this:

> Here, u-boot is built without referencing arm-trusted-firmware, and then armTrustedFirmwareA3700 would bring u-boot.bin, wtmi_app.bin, and a few other files that I haven't understood how to bring into Nix

So you've not produced a firmware that you have used on your hardware yet, right?

No worries if you haven't yet, we can look and sort this out! I just wanted to signal it accurately in the PR :)

File: pkgs/misc/mox-boot-builder/default.nix
Comment (samueldr) [inline pkgs/misc/mox-boot-builder/default.nix:9]: This name is inaccurate. while this repo is named `mox-boot-builder`, what it actually builds is a firmware that adds support for the hardware random number generator.

 - https://github.com/ARM-software/arm-trusted-firmware/blob/5a53c6c6679e189eda7233232570958263332883/docs/plat/marvell/armada/build.rst?plain=1#L273-L295

I'm not sure what really it should be named. `armada-3720-secure-firmware` wouldn't be bad, but that name would *also* be somewhat wrong, since it's more like `nic-cz-armada-3720-hardware-rng-firmware`. Though that would be a mouthful.

Anyone has suggestions?

**Comment by samueldr** [comment]
And here's a commit describing the changes for the platform:

 - https://github.com/samueldr/nixpkgs/commit/2d3e83a9b83422d57875b0b4af11bc5a2d6f0f5c

I have tested that it builds either through cross-compilation or natively on `aarch64-linux`.

File: pkgs/misc/uboot/default.nix
Comment (ManoftheSea) [reply pkgs/misc/uboot/default.nix:205]: Alright.  The u-boot code does work on older revisions, just that I contributed a patch that fixes overwriting environment variables that only made it into 2023.01.  Would you like me to add a revision to remove the version?

File: pkgs/misc/mox-boot-builder/default.nix
Comment (ManoftheSea) [reply pkgs/misc/mox-boot-builder/default.nix:42]: I have asked Pali, one of the developers, to provide more information.  Hopefully, that's in the form of a license text in the repo.

**Comment by ManoftheSea** [comment]
To clarify some asked questions:
"mox-boot-builder" builds BL32, code that runs in the secure world of an armv8 processor, for the armada 3720.  This also enables a hardware RNG that the vendor firmware does not expose.
I have produced and used firmware for both v5 and v7 versions of the espressobin from u-boot 2023.01, TF-A 2.8, wtmi_app.bin 2022.06.11-7-g0290b2c, marvell DDR and Marvell A3700-utils; I have not done so in the declarative Nix way. TF-A seems to want to reach into the DDR and utils repos and call git commands on them, rather than just receive files.  Maybe this can be resolved by including a patch on the Makefile in Nix?  Open to suggestion
I am attempting to work with the turris repo developers to tag their code with a license.

File: pkgs/misc/mox-boot-builder/default.nix
Comment (ManoftheSea) [reply pkgs/misc/mox-boot-builder/default.nix]: As far as I am aware, there are two versions of BL32 software for armada 3700 (a3k) chips: the initial marvell code which does nothing more than programs some efuses, and the turris code which provides the hwrng.  Turris can sign their firmware for their Mox boards; here we use it unsigned.  I have offered "nic-cz-armada-bl32-unsigned-firmware".

**Comment by abathur** [comment]
@ManoftheSea Did you hear anything back on the upstream questions here, and are you still interested in pursuing this?

**Comment by ManoftheSea** [comment]
I have abandoned this.  Sorry.

**Comment by abathur** [comment]
> Sorry.

@ManoftheSea No worries--thanks for the quick reply. I'll go ahead and close it, but feel free to mention me if you change your mind and want it reopened.

**Pull request closed without merging**
